proxy/http/remap/NextHopConfig.cc

```diff
@@ -0,0 +1,387 @@
+
+#define BUFSIZE 8192
+
+ts::Errata
+NextHopConfig::loadConfig(const char *fileName)
+{
+  std::stringstream doc;
+  std::unordered_set<std::string> include_once;
+
+  try {
+    loadFile(fileName, doc, include_once);
+    YAML::Node node = YAML::Load(doc);
+    config          = node.as<NextHopStrategyConfig>();
+
+  } catch (std::exception &ex) {
+    errata.push(ts::Errata::Message(1, 1, ex.what()));
+    return errata;
+  }
+  return errata;
+}
+
+void
+NextHopConfig::loadFile(const std::string fileName, std::stringstream &doc, std::unordered_set<std::string> &include_once)
+{
+  const char *sep = " \t";
+  char *tok, *last;
+  char line[BUFSIZE];
+
+  std::ifstream file(fileName);
+  if (file.is_open()) {
+    while (file.getline(line, BUFSIZE - 1)) {
+      if (line[0] == '#') {
+        tok = strtok_r(line, sep, &last);
+        if (tok != nullptr && strcmp(tok, "#include") == 0) {
+          std::string f = strtok_r(nullptr, sep, &last);
+          if (include_once.find(f) == include_once.end()) {
+            include_once.insert(f);
+            try {
+              loadFile(f, doc, include_once);
+            } catch (std::exception &ex) {
+              throw;
+            }
+          }
+        }
+      } else {
+        doc << line << "\n";
+      }
+    }
+    file.close();
+  } else {
+    throw std::runtime_error("unable to load '" + fileName + "'");
+  }
+}
+
+namespace YAML
+{
+// decodes a 'NextHopStrategyConfig' type.
+template <> struct convert<NextHopStrategyConfig> {
+  static bool
+  decode(const Node &node, NextHopStrategyConfig &cfg)
+  {
+    YAML::Node strategy = node;
+    try {
+      strategy = node[NH_strategy];
+    } catch (std::exception &ex) {
+      throw std::runtime_error("the required 'strategy' node does not exist in the yaml document.");
+    }
+    if (strategy.Type() == YAML::NodeType::Map) {
+      // verify keys
+      for (const_iterator it = strategy.begin(); it != strategy.end(); it++) {
+        if (std::none_of(valid_strategy_keys.begin(), valid_strategy_keys.end(),
+                         [it](std::string s) { return s == it->first.as<std::string>(); })) {
+          throw std::invalid_argument("unsupported strategy key: " + it->first.as<std::string>());
+        }
+      }
+      std::string policy;
+      try {
+        policy = strategy[NH_policy].as<std::string>();
+      } catch (std::exception &ex) {
+        throw std::invalid_argument("required 'policy' field was not found");
+      }
+      cfg.policy = static_cast<NextHopSelectionPolicy>(POLICY_DESCRIPTOR.get(policy));
+      if (cfg.policy < 0) {
+        throw std::invalid_argument("unknown policy value '" + policy + "'");
+      }
+
+      // parse the groups list.
+      YAML::Node groups;
+      try {
+        groups = node[NH_groups];
+      } catch (std::exception &ex) {
+        throw std::invalid_argument("the required 'groups' node is not defined in the strategy.");
+      }
+      if (!groups.IsNull()) {
```

Comment:
Why check for Null, instead of just checking for Sequence? Either it's a Sequence (valid) or not.

Reply:
right you are.